### DISCUSSION TITLE
Can I integrate Arduino library ACS712-arduino-master into Tasmota?

### DISCUSSION THREAD

**Discussion by ojoojoo** [arendst/Tasmota#18172] [Q&A]
When I connect the ACS712 30A sensor on A0 in Tasmota I do not get valid values when I configure ADC CT Power on GPIO17 (A0)
![2023-03-12 18_18_50-Tasmota - Main Menu](https://user-images.githubusercontent.com/56216880/224561178-8e185855-d4cc-4f11-ba12-c8adeae7a9ff.png)

But when I load this Arduino code:
/*ACS712 Current Sensor Demo
   By miliohm.com
*/

#include "ACS712.h"
#include <Wire.h>
#include <LiquidCrystal_I2C.h>

LiquidCrystal_I2C lcd(0x3F, 16, 2);

ACS712 sensor(ACS712_30A, A0);
//ACS712_20A for 20 Amp type
//ACS712_30A for 30 Amp type

void setup() {
  Serial.begin(9600);
  sensor.calibrate();
  lcd.init();
  lcd.backlight();
  lcd.setCursor(0, 0);
  lcd.print("ACS712 Demo");
  lcd.print("By miliohm.com");
}

void loop() {
  float I = sensor.getCurrentAC();
  if (I < 0.09) {
    I = 0;
  }
  lcd.setCursor(0, 0);
  lcd.print("Current : ");
  lcd.print(I);
  lcd.print(" A");
  delay(300);
}

It is working correctly and I can read the correct current value on the LCD display.
Its using the ACS712-arduino-master library.

Can I easily integrate this in Tasmota?

**Comment by barbudor** [comment]
"CT Power" means current transformer.
This mode is intended to support current transforrmers such as SCT-013

You could integrate ACS712 support into the `xsns_02_analog.ino` driver for the analog input as an additional mode for the ADC Input

**Reply by sfromis** [reply]
If the analog value you get can be linearly mapped to current, you need no special ACS712 support, as this would be workable with the regular config for ADC Range, where you use the command `AdcParam` to config two points on the line.

**Reply by barbudor** [reply]
it requires an integration

https://github.com/rkoptev/ACS712-arduino/blob/0f426a6fefd57a61680ba6e2a824cee8f33ff89f/src/ACS712.cpp#L46-L61

**Reply by ojoojoo** [reply]
> "CT Power" means current transformer. This mode is intended to support current transforrmers such as SCT-013
> 
> You could integrate ACS712 support into the `xsns_02_analog.ino` driver for the analog input as an additional mode for the ADC Input

Thank you for your answer, do you have to configure the SCT-013 somewhere, because you can buy them for example in the"5A 1V", "15A 1V", "60A 1V" or "100A 1V" version? Which one can you recommend?

**Reply by barbudor** [reply]
Yes, as documented here : https://tasmota.github.io/docs/ADC/#commands
Using the command "AdcParams 7" 

Advantage of the CT is that you have total isolation between your circuit and circuit-under-measure
ACS712 is supposed to do the same as it is based on hall effect. But it's a tiny chip, if you put your finger at the wrong location, you will get zapped ⚡ ⚡ 

Advantage of ACS is that probably you don't need much to connect to an ADC input

With CT you need need a bit of circuit adaptation : 
https://github.com/arendst/Tasmota/blob/bda32739a8d55f970f537435c12c4e36bb4f7926/tasmota/tasmota_xsns_sensor/xsns_02_analog.ino#L68-L73

**Reply by ojoojoo** [reply]
> If the analog value you get can be linearly mapped to current, you need no special ACS712 support, as this would be workable with the regular config for ADC Range, where you use the command `AdcParam` to config two points on the line.

Thanks for the answer, but what I'm actually looking for is the Voltage, Current, Active Power and Energy Total from the ADC CT Power module, I don't see how to get this from a range?

**Reply by Sayed-afridy** [reply]
how can i add this integration?

**Reply by barbudor** [reply]
Have you read my answers above?

**Reply by Sayed-afridy** [reply]
yes but im just not sure where to add the code of this link [ https://github.com/rkoptev/ACS712-arduino/blob/0f426a6fefd57a61680ba6e2a824cee8f33ff89f/src/ACS712.cpp#L46-L61 ] inside the xsns_02_analog.ino . i have added the code inside the xsns_02_analog.ino but its giving me error while building

**Reply by Sayed-afridy** [reply]
actually this link is not working

**Comment by ojoojoo** [comment]
How do I configure this using the serial console? 

CT Power | Voltage 230 VCurrent %value APower %value WEnergy Total %value kWh | {"Energy":%value,"Power":%value,"Voltage":230,"Current":%value}

**Reply by barbudor** [reply]
See above, the configuration command is `AdcParams 7`

**Reply by ojoojoo** [reply]
Thanks a lot for the info, I have a lot of reading to do, but you helped me in the right direction!

**Reply by barbudor** [reply]
You will get that when you have correctly configured the GPIO as ADC_CT and configured AdcParams 7

![image](https://user-images.githubusercontent.com/2996042/224807372-5085aff9-a6b9-4396-9b59-c90efb25e885.png)

**Reply by ojoojoo** [reply]
Thanks for the example "AdcParam 7, 406, 3282, 0.012", this was just what I was looking for, I'm a newbie and need to find my way into this, thanks again.

**Reply by barbudor** [reply]
This  example is in the documentation I already linked above

**Comment by ojoojoo** [comment]
Sorry for asking again, but is there a way to get the same results for the ACS712 using the ADC CT Power module to get the same results when using the Arduino library https://github.com/rkoptev/ACS712-arduino/blob/0f426a6fefd57a61680ba6e2a824cee8f33ff89f/src/ACS712.cpp#L46-L61 ?

#ifdef ESP8266
#define ANALOG_RESOLUTION             10               // 12 = 4095, 11 = 2047, 10 = 1023
#define ANALOG_RANGE                  1023             // 4095 = 12, 2047 = 11, 1023 = 10
#define ANALOG_PERCENT                10               // backward compatible div10 range
#endif  // ESP8266

// CT Based Apparrent Power Measurement Parameters
// 3V3 --- R1 ----v--- R1 --- Gnd
//                |
//         CT+   CT-
//          |
//        ADC0
// Default settings for a 20A/1V Current Transformer.
// Analog peak to peak range is measured and converted to RMS current using ANALOG_CT_MULTIPLIER
#define ANALOG_CT_FLAGS               0                // (uint32_t) reserved for possible future use
#define ANALOG_CT_MULTIPLIER          2146             // (uint32_t) Multiplier*100000 to convert raw ADC peak to peak range 0..ANALOG_RANGE to RMS current in Amps. Value of 100000 corresponds to 1
#define ANALOG_CT_VOLTAGE             2300             // (int) Convert current in Amps to apparrent power in Watts using voltage in Volts*10. Value of 2200 corresponds to 220V

**Reply by barbudor** [reply]
The CT mode is meant to be used with a CT module

If you want to use ACS712, you need to implement a new mode.
You can get inspiration from the CT module but it has to be a new mode
You can't use the CT mode with the ACS712

**Reply by ojoojoo** [reply]
Sorry for asking again, but when I connect a CT coil with default ratio 20A/1V to the analog A0 I get half the value with the default settings:
AdcParam 7, 0, 2146, 0.23

When I configure this:
AdcParam 7, 0, 4292, 0.23

It's reading the correct values, is the Tasmota documenation wrong or am I missing something?

7 = CT POWER parameter adjustments:
<param1> = ANALOG_CT_FLAGS (default 0 for a non-invasive current sensor). When value is >0 it sets the adcLow value as base for the measurement via OpAmp differential amplifier.
**<param2> = ANALOG_CT_MULTIPLIER ( 2146 = Default settings for a (AC) 20A/1V Current Transformer.) multiplier*100000 to convert raw** ADC peak to peak range 0..1023 to RMS current in Amps. Value of 100000 corresponds to 1
<param3> = ANALOG_CT_VOLTAGE (default 2300) to convert current in Amps to apparent power in Watts using voltage in Volts*10. Value of 2200 corresponds to AC220V. For DC it's Volt/1000. Eg. 12VDC = 0.012.
AdcParam 7,406,3282,0.012

Also when no current is passing it always says this:
![image](https://user-images.githubusercontent.com/56216880/229286862-459377e3-839f-4558-a374-a3ce11e9e612.png)

So without any power it's displaying already 10w of usage, is there any way to tune this?

This is diplayed when a current is passing:
![image](https://user-images.githubusercontent.com/56216880/229286935-7f06185e-e878-42c8-97a8-c2fd9e0e49bd.png)

**Reply by barbudor** [reply]
Which ESP board are you using ?
Have you connected your CT accordingly to the schematic in the source code ? What is your values for R1 ?

If you set the input as ADC_Input, what does the values looks like when no current is passing ? (or CT in the air without a conductor going through). Are the values stable and constant and close to 512 ?

**Reply by ojoojoo** [reply]
I use these param: AdcParam 7, 0, 2146, 0.23 for a 20A/1V CT coil

Which ESP board are you using ? ESP8266
Have you connected your CT accordingly to the schematic in the source code ? **Yes
// CT Based Apparrent Power Measurement Parameters
// 3V3 --- R1 ----v--- R1 --- Gnd
// ...........................|
//         CT+       ................CT-
//          |
//        ADC0**

What is your values for R1 ? **I used 2x 100kΩ resistors**

If you set the input as ADC_Input, what does the values looks like when no current is passing ? (or CT in the air without a conductor going through). Are the values stable and constant and close to 512 ?
![image](https://user-images.githubusercontent.com/56216880/229364201-153c5c1b-5463-4ad1-ba8f-555874be9269.png)

It's around 452-455 without load (in the air)

**Reply by barbudor** [reply]
> Which ESP board are you using ? ESP8266

ESP8266 is not a board
Are you using a NodeMCU or a mini-D1 ?


I believe you have too much noise and some capacitors may be helpful. Also 100K might be too high
I would recommend testing this:

![image](https://user-images.githubusercontent.com/2996042/229374235-92233bac-141b-4486-9735-6685f9b7f363.png)

I will try myself in the coming days but not today

**Reply by ojoojoo** [reply]
Your right I'm using a NodeMCU board with a 8266 on it.
First I need to get this working properly and than my ultimate goal is to use hall sensors for current measuring around a building. Something like a ABB Component System pro M compact - CMS-121PS current sensor.

I have a friend who is a software developer, I think he can help me to realize this in the Tasmota code as a new mode.

If you can help me test this, it would be great, thanks for the help so far!

**Reply by barbudor** [reply]
I have a hard time to follow you
You first talked about ACS712 then went to CT and now you are talking about those CMS-121PS sensors
Looks like a nice product but not analog at all. It seems to use a digital proprietary bus (CMS-bus) to communicate with a central unit.
The electrical interface is RS485 but do you have the specs of the electrical interface and the protocol ?

**Reply by ojoojoo** [reply]
Sorry for the confusion, I'm working on several ideas that have a similarity with each other.
First I want to get a proper reading for the CT coil on the Tasmota.
Second, I want to make an new integration with a ACS712 as a new function in Tasmota (for some reason nobody ever made this)
Thirth, the ultimate goal is to use a common hall sensor for example a 3144 to measure current

So 3 things, but first the CT coil working the way it supposed to work.

I build your scheme with slightly different capacitors, I used a 117nF non polar capacitor (no ceramic) instead of a 110nF and a 10μF polarized capacitor instead of the 1μF, but the readings doesn't improve see below:
![image](https://user-images.githubusercontent.com/56216880/230040831-4585f506-7d60-4746-a546-36cde67c8336.png)

![b](https://user-images.githubusercontent.com/56216880/230041236-c7f75c41-69af-4cb2-8aaf-cf2ccdc58299.jpg)
![a](https://user-images.githubusercontent.com/56216880/230041245-e56d87d0-d043-4253-8b07-562b296b8355.jpg)

Should it matter this much to use different capacitors, I don't have the corrt ones in stock for now?

Thanks for the help so far, regards Orson

**Reply by barbudor** [reply]
Hi
Capacitors values are not critical but a clean and short circuit is very important
I did the following using 68nF and 10µF capacitors on a small proto board, all soldered, no dupont wires, as compact as possible

![image](https://user-images.githubusercontent.com/2996042/230735135-0e756d12-5027-436f-a8a9-3254920c7e58.png)

The noise is now very limited to 2 LSB (the ADC value is around 545 +/- 1)
My CT is 50A/1V 
Note that it is important to use a model Current => Voltage and not a model Current => Current
I can't see for sure on your photo if htis is your case

I made 2 little changes on the driver:
- Better calculation, should be more accurate
- Support of an additional 4th parameter which allow to define a current threshold

Using my 50A/1V CT I have calibrated it using known loads : an old 40W filament buld and a hair-dryer with 2 speed
My calibration led to `AdcParam 7, 0, 5700, 0.245`
The noise level of 2 LSB translate into a no-load current of 0.057A and 13,9W

So with the new 4th parameter, I can set `AdcParam 7, 0, 5700, 0.245, 0.060` which set the minimum current to 0.060 A
I still have correct values when testing my 40W buld or the hair dryer but when no current goes through, it is forced to 0A/0W

So as a summary:
- You need a very clean circuit to reduce the noise a much as possible
- Don't expect an exact formula, you will have to calibrate using a known power load to find the correct value by trial
- Once calibration is done, look at the value when no current is going through and apply a slighly higher value as the 4th parameter to AdcParam

I will push this change to Tasmota
EDIT: PR has been merged. Get latest binaries

**Reply by barbudor** [reply]
PR https://github.com/arendst/Tasmota/pull/18369 has been placed
You can also wait for merge and new dev binaries to be available

**Reply by barbudor** [reply]
PR has been merged

**Reply by ojoojoo** [reply]
Hello,

I rebuild this using adifferent NodeMCU and I flashed your firmware version to it, and yes it's quite an improvement, the analog reading is rocksteady at 540 when it doesn't have a load. In this setting I use a different CT with a 10A/1V rating.
This is something we can absolutely use.

I'm using these settings: AdcParam 7, 0, 1140, 0.23,  0.060

![image](https://user-images.githubusercontent.com/56216880/230778944-36602d33-6486-4f5e-952c-2edee3e4dd01.png)

When I test this with a 2000W paint blower it matches exactly with the output of the multimeter (Yes, the wiring looks sketchy and dangerous, and it probably is, it's living on the edge :))

![image](https://user-images.githubusercontent.com/56216880/230779428-543b3efe-9c3d-445b-a272-9d94c29e76b1.png)
![df](https://user-images.githubusercontent.com/56216880/230779555-af9278fe-aa59-47f5-a70d-cd14af6d5d87.jpg)

Thanks for the amazing work and help!

Is there also a way to measure the Voltage and the current to the ESP? Now the voltage is set at 230V, but this changes al lot nowadays with all the solar panels around>

**Reply by barbudor** [reply]
Not with a CT which is only picking thecurrent going through the wire
Of course I would not advise you of direct measurment

An possible safe idea in the same way of the CT is to power your circuit not with a modern power supply but with a good all transformer (like 220V to 6V) + bridge rectifier and mesuring the peak voltage after the bridge rectifier. With a bit of calculation it would probably give you a good estimation.

Now the simplest way is to use an standard Tasmotizable product just like a plug or a sonoff/shelly and that will give you the voltage

**Comment by barbudor** [comment]
Docs have been updated with the new 4th optional parameter and the recommended circuit 
https://tasmota.github.io/docs/ADC/#adc_ct-circuit

**Comment by barbudor** [comment]
@ojoojoo 
I think you can use ACS712 using ADC_CT the following way:

![image](https://user-images.githubusercontent.com/2996042/230942798-0a56d7cb-5be2-4456-ba3a-1fc73ef8aa74.png)

With ACS712-05A => R1 = 27 kohm
With ACS712-20A and -30A => R1 = 180 kohm
You need exact value

You will have to find the proper calibration value for AdcParam but that should work

I will see if I have time to test it myself

Note that the above diagram is for Nodemcu/mini-D1 ESP8266 only
For ESP32, a different circuit would be required

**Reply by ojoojoo** [reply]
Hi,

I tested this and it's kind of working, but I have to set the last parameter value to 2A wich is kind of high, because this also means that any current draw just under 2A is not registered.
Im using the ACS712-30A for this and found that these parameter values are reasonably well: AdcParam 7, 0, 2200, 0.23, 2.0

When I set the last parameter value at 0.0 and no current passes this is my reading:
![image](https://user-images.githubusercontent.com/56216880/230971595-047bacb6-0a46-481d-9b38-dcbc14332e81.png)

I didn't have a 180K resistor, thats why I used 100K and 4x 20K in series which measures 177,8K Ohm.
You said that this should be exact, but that's difficult, how did you calculate this to 180K Ohms?

I get this when I set it to ADC input:
![image](https://user-images.githubusercontent.com/56216880/230972406-16bae7f8-dd9e-4fd7-ace8-a988cf9613b9.png)

When this is working properly I can finally experiment with another hall sensor by rebuiling the circuit of the ACS712 using a 3144 hall sensor instead of the closed hall sensor on the ACS712.

Thanks for all the great work so far!

**Reply by barbudor** [reply]
I will try to test with ACS712 in the next couple of days

The calculation of the resistors comes from circuit considering that:
- ACS712-5A provides an output equal to 2.5V +/- 0.950V => Vmax = 3.45V
- ACS712-20A and -30A provide an output equal to 2.5V +/- 2.08V => Vmax = 4.58V

But the input of the ESP8266 ADC input should never go over 1V

R1 should be calculated so that:
VMax * (100K / (R1 + 220K + 100K)) <= 1V
So R1 >= 100K * Vmax - 320K
With ACS712-5A we get R1 >= 25K. So I choose the next valid value which is 27K
With -20/-30 we get R1 >= 138K. So actually 150K would be ok (not sure why I got 180K in first round).


A3144 is a generic hall sensor
How are you expecting to transfor that into a current sensor ? Math are going to be odd.
If you want a current sensor to put above your wire, look at those Allegro devices such as the ACS37610/37612 https://www.allegromicro.com/en/products/sense/current-sensor-ics/sip-package-zero-to-thousand-amp-sensor-ics
But I've been told this is also complicated to use (see the section of the datasheet on how this has to be mounted)

You are really going into every direction :exploding_head:

**Reply by ojoojoo** [reply]
Thank you for the explanation, it now totally makes sense to me.
It looks like I'm going in every direction, but there is a certain logic to it to me. I'm exploring fields in electronics I have no experience with and that makes it interesting. You might be right about the hall sensor, but when I don't explore this option I'll never know.

The The ACS37612 measures currents from 200A to >1000A flowing through a PCB or a busbar with a typical accuracy of 1%, and the ACS37610 provides a highly accurate method of measuring 100 A to >4000 A.
This is way out of the range that I want to use it for, my goal is to measure from 0-20A AC max.

The A3144 gives an output Voltage proportional to applied magnetic field, thus linear, according to the data sheet.
So maybe I'm stubborn, but I can't see why it's impossible to use this. The only problem could be that the resolution would be very low and therfore yields inaccurate measurements.

ABB has these type of open core current sensors, these are very small and accurate (https://new.abb.com/products/nl/2CCA880211R0001/cms-121ps)
The only way I can imagine is that they also work with the hall sensor technology.

So finally you know my ultimate goal, and that is to build this type of sensor myself, but than for use with Tasmota and a ESP.
Thanks for being so patient with me anyways!

**Reply by Sayed-afridy** [reply]
what will be the value for esp32?

**Comment by Sayed-afridy** [comment]
> Have you read my answers above?

yes but im just not sure where to add the code of this link [ https://github.com/rkoptev/ACS712-arduino/blob/0f426a6fefd57a61680ba6e2a824cee8f33ff89f/src/ACS712.cpp#L46-L61 ] inside the xsns_02_analog.ino . i have added the code inside the xsns_02_analog.ino but its giving me error while building

**Reply by barbudor** [reply]
You don't need to add any code to Tasmota
The proper mode for Tasmota to handle such sensor is to use the ADC input in CT mode. This is already providing the integrations.
Then you will need to adjust `AdcParam 7` to reach the proper values

The ACS712 is measuring the current and converting it to a voltage
The exact AdcParams depends on your ACS712 model and of your ESP board


By the way, the URL you gave goes nowhere. There is no such repository.

**Reply by Sayed-afridy** [reply]
the problem i am facing is its showing consumption even if there is no load.
 about the link : yes you gave this link earlier but its not available anymore.

**Reply by abhishekakt1** [reply]
@barbudor Do you know the AdcParams I use in for ACS712 30Amp sensor?
